### ISSUE TITLE
Encoding problem when sending mails with HTML body

### ISSUE THREAD

**Issue by bwolff** [microsoftgraph/msgraph-sdk-java#779]
Dear Developers,

this issue here is a follow-up on a question that I already raised here: https://docs.microsoft.com/en-us/answers/questions/392676/wrong-encoding-when-sending-mails-with-html-body-u.html

### Context

I'm using the Graph-API Java SDK in version 3.6.0 to send mails with HTML body from my platform (Windows 10, 64bit, Java OpenJDK 16) using a shared mailbox and an application account (authenticated with app credentials). However, I'm experiencing character encoding issues.

### Expected behaviour

Sending mails with an HTML body that contains non-ASCII characters should use the correct encoding and display the characters correctly in the mail client.

### Actual behavior

Sending mails with an HTML body that contains non-ASCII characters results in a wrong encoding headers in the generated mails and the characters appear garbled in the mail client. Below are example screenshot of how a test mail display in my Outlook client:

![encoding_issue_1](https://user-images.githubusercontent.com/1532631/119645964-e24e2300-be1e-11eb-92b9-e90b0263e804.png)

Using the "show source" function, the HTML body reveals that an obviously wrong encoding is chosen by the Exchange server (or whatever generates the mails and decides on the encoding):

![encoding_issue_2](https://user-images.githubusercontent.com/1532631/119646102-11649480-be1f-11eb-9c58-ca1e31787d07.png)

The reason why I'm creating this issue here for the Java SDK is that I cannot reproduce this using Postman. With Postman I'm getting a different behaviour using the same API endpoints, so I assume it's an issue with either the Graph-API Java SDK or my platform.

Sending the same mail content using a raw POST request in Postman, the characters to not appear garbled and the encoding header is now a different one:

![encoding_issue_3](https://user-images.githubusercontent.com/1532631/119647480-9c925a00-be20-11eb-80a3-7c155381cece.png)

![encoding_issue_4](https://user-images.githubusercontent.com/1532631/119647489-9ef4b400-be20-11eb-9fe9-d3b78be7e3b8.png)

The encoding is now ISO-8859-1, which is still not optimal, since it should be UTF-8, **but it seems that there is no possibility to control the encoding of the HTML body with the client API.** According to my tests, the mailing server seems to choose an encoding that fits the provided body, which doesn't seem to work in this case.

Btw, I'm experiencing the same problem when creating events with HTML bodies.

See below for steps how to reproduce this.

### Steps to reproduce the behavior

I have attached a small Gradle projects with code that should help to reproduce the issue: [encoding-issue-demo.zip](https://github.com/microsoftgraph/msgraph-sdk-java/files/6546052/encoding-issue-demo.zip)

1. Unzip the file into a folder of your choice
2. Adjust the code in `encoding.issue.demo.Main` for the test recipient, the mailbox to send from, and the authentication information for your test environment.
3. Send a test mail by running `gradle run`
4. Check if the resulting mail has the described encoding issues as described above.

You can also try to enforce the platform encoding in the `build.gradle` file by commenting in the `file.encoding=utf-8` start argument, but it didn't change the beahavior in my case.

As I mentioned above, I can't reproduce this issue using a raw POST request. I'm sending the following payload using Postman:

```json
POST https://graph.microsoft.com/v1.0/users/{{MailboxName}}/sendMail

{
    "message": {
        "subject": "Encoding Issue Test-Mail from Postman",
        "body": {
            "contentType": "HTML",
            "content": "<p>Test-Mail sent with Postman.</p><p>öäü#+ß^°123§%&/()=?`´#'*~+-<>|,;.:-_\"áàúù\\{}[]âûôêéè</p>"
        },
        "toRecipients": [
            {
                "emailAddress": {
                    "address": "<insert recipient address>"
                }
            }
        ]
    }
}
```

There have been encoding issues in the past, like #95 , but this seems to be a different problem.

Can the described behavior be reproduced? If not, do you maybe have an idea if this could be a problem with the configuration of the AD tenant? The fact that the behavior is different between Java and Postman is, however, an indicator that it could be a problem with the SDK here.

I hope you can help to shed some light on this issue. :)

Best regards,
Ben

 
[AB#9606](https://microsoftgraph.visualstudio.com/0985d294-5762-4bc2-a565-161ef349ca3e/_workitems/edit/9606)

**Comment by baywet** [comment]
Hi @bwolff 
Thanks for reaching out and for the detailed explanation + repro app!
I've tried your app locally, and I noticed something interesting:
- When I run it with gradle run, it misbehaves like you described
- When I run it with VS code F5, it runs properly

It sounds like it's a JVM configuration issue of some sort but I can't pinpoint it. It'd be interesting to see if you could keep tweaking the JVM configuration when running through gradle to see if you can get it to work.

**Comment by bwolff** [comment]
Hi @baywet,

Thanks for the quick return on this! At least we are one step further in the sense that you can reproduce the described issue ;).

Now it becomes really interesting: when I run the app in Eclipse via "Run as..." (basically the F5 as well), and not via "Gradle run", it also behaves as expected and the problem does not surface. Therefore it really seems to be an issue with how Gradle spools up the JVM.

I will continue to investigate the issue to see if I can pinpoint the problem with Gradle. While probably not being an issue with the Graph-API, it might be worth to file this under "known issues". I'll come back with what I can find out.

Cheers,
Ben

**Comment by bwolff** [comment]
Ok, I spent a bit more time on this and found the culprit. It was kind of a tough nut to crack ;). At the end it was indeed a platform encoding issue. The Gradle build was using the Windows-1252 encoding at compile time to create the .class files, and that's what ultimately caused the observed problems.

### Anatomy of the problem

- When running a Gradle task (e.g.  `run`, which also triggers `build`and `classes`), it first starts a daemon in the background. If not explicity configured in the Gradle environment, either at the OS user or the project level, the daemon falls back to use the default platform encoding, which is Windows-1252 on my Win10 machine.
- When the initial compilation of the Java file is done, Gradle uses this default platform encoding to generate the .class files. It seems that at this point, the character encoding issue is already present within the class file.
- Now when running the code, specifying the "-Dfile.encoding=utf-8" didn't help, because the encoding problem was already present in the class file. Providing this start parameter for `gradle run` only changes the default charset and encoding of the running JVM that runs the code. So although the running code prints out utf-8 as the encoding when executed, the hardcoded strings in the .class have probably stored with windows-1252 encoding, which is incompatible with utf-8 in some non-ASCII characters.

### Possible implication

What scares me a bit is the fact that I can also reproduce this problem even if I specify windows-1252 encoding at build and at runtime. I think this is a result of fixing #95. Since retrieval of the bytes of the strings was fixed to utf-8, my hypothesis is that if the Java sources are compiled with a different file encoding than utf-8, any incompatibly encoded special characters in hardcoded strings in the sources will produce the observed behaviour when talking to the Graph API.

This problem probably only surfaces when non-ASCII characters are used, so it might not occur in many usages. I leave it to you to judge the severity of this and to reproduce the described issue. It should the default to use utf-8 nowadays, but if you miss this details, you may end up spending quite some time troubleshooting this (like I did ;)).

### Solution

To fix the problem at hand it should be ensured that the sources are built/compiled with utf-8 file encoding. For Gradle, this can be configured by creating a file `gradle.properties` with the content `org.gradle.jvmargs=-Dfile.encoding=utf-8` and placing it in the project folder (project level) or by placing it under `<user home>/.gradle/` (local environment level).

As a side note: what made the problem a bit more tricky to troubleshoot was the fact that Gradle did not recompile the class files after I changed this configuration in my environment. Thus it was still reproducing the same behaviour, although I changed the platform encoding already. I was not far from going crazy, but then remember to run a clean-build and then the issue was fixed ;).

HTH,
Ben

**Comment by baywet** [comment]
Hi @bwolff 
Thank you very much for the detailed investigation and explanation here!

From what I read, there's not much to be done at this point in the SDK itself unless I'm missing something. It's already using the proper encoding to get the string as a byte array, and the strings are not in the SDK's classes. We could do two things to be future proof however:
- add this configuration to the `gradle.properties` just in case the strings might contain UTF8 information in the future.
- add some kind of known issue for consumers to be aware of in the docs so people running into this issue know how to fix it.

Am I missing anything?

**Comment by bwolff** [comment]
Hi @baywet

I agree that there is nothing to be done in the SDK at this point. Redarging your suggestions:

* Adding the `-Dfile.encoding=utf-8` flag to the SDK project's `gradle.properties`: if it is expected that the development environments treat the source and ressource files with utf-8 encoding, I think that would be a good thing to do. It would make the project itself a bit more robust against misconfigured local environments. However, since this controls the file encoding setting for all Gradle commands executed within the project, I'm not sure if this could lead to unexpected side-effects. Maybe you could ask some colleagues if they think that this could be a problem?
* Known issue: imho, this could go into a "known issue" or the "how to use" section of the SDK docs. If you could point me to the respective files at let me know where to add it, I could create a pull-request to add this.

Cheers
Ben

**Comment by baywet** [comment]
Hi @bwolff,
The azure folks [are using maven](https://github.com/Azure/azure-sdk-for-java/tree/master/sdk/identity) so I don't think they ran into the problem at all.
We don't have a placeholder for that information today, but arguably it's some setup one needs to go through when taking a dependency on the Graph SDK with gradle with international strings. 
So maybe this could be added as a note (with the `> NOTE: ` prefix) right under the [Groovy block](https://github.com/microsoftgraph/msgraph-sdk-java#11-install-via-gradle)?

Before we get into this though, I had a last idea, what if you put the string in a resource instead and pull it from the resource files? I'm not sure whether resources end up in class files (and therefore would have the same issue) or as separate files. Do you think it's worth a try? Would you mind giving it a try? https://docs.oracle.com/javase/tutorial/i18n/resbundle/concept.html

**Comment by bwolff** [comment]
After thinking about it a bit more, I think the best solution is no simply close this issue here without any further action.

The problem I experienced is not really specific to Gradle/Groovy, it's caused by a wrong platform encoding at compile-time. However, two things must come together to cause this issue:

- a byte-incompatible (for special characters) platform encoding to utf-8 at compile time;
- using hardcoded strings with non-ASCII special characters in the Java source;

I used a small Gradle project to explore the Graph API and playing around with it and that when I encountered the issue. In the production system, the payload would not come from static strings in the sources, but rather from e-mail templates, resources files (as you mention), and so on.

Long story short, I think it's very rare that this issue will be experienced by someone else and if so, probably during prototyping around with a small poc project.

I think we're golden by just closing this issue. Since we have this issue here now on GitHub and I already raised the question on Microsoft Questions, affected people should be able to find the solution by "binging" (I should probably use this termin to search the interwebs in a Microsoft context :D ) for it.

**Comment by baywet** [comment]
Thanks for taking the time to go through this issue thoroughly! And yes I agree with the general conclusion, localized strings should either come as data from a service or come from a localization resource but not be hardcoded in the source code.
Plus, this issue will serve as a reference point for anybody encountering a similar issue.
Closing.
